File: go/cmd/network-rtt/tcp.go
package main

import (
	"bytes"
	"fmt"
	"io"
	"net"
	"time"
)

// measureTCP starts an in-process TCP echo server on loopback, connects one
// client with TCP_NODELAY set, runs warmup then cfg.Iterations strict
// ping-pong round trips, and returns the per-round-trip elapsed nanoseconds.
func measureTCP(cfg Config) ([]int64, error) {
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		return nil, fmt.Errorf("tcp: listen: %w", err)
	}
	defer ln.Close()

	srvErr := make(chan error, 1)
	go tcpEchoServer(ln, cfg.PayloadBytes, srvErr)

	conn, err := net.Dial("tcp", ln.Addr().String())
	if err != nil {
		return nil, fmt.Errorf("tcp: dial: %w", err)
	}
	defer conn.Close()

	tcpConn, ok := conn.(*net.TCPConn)
	if !ok {
		return nil, fmt.Errorf("tcp: expected *net.TCPConn, got %T", conn)
	}
	if err := tcpConn.SetNoDelay(true); err != nil {
		return nil, fmt.Errorf("tcp: set nodelay: %w", err)
	}

	send := make([]byte, cfg.PayloadBytes)
	for i := range send {
		send[i] = byte(i)
	}
	recv := make([]byte, cfg.PayloadBytes)

	roundTrip := func() error {
		if _, err := tcpConn.Write(send); err != nil {
			return fmt.Errorf("tcp: write: %w", err)
		}
		if _, err := io.ReadFull(tcpConn, recv); err != nil {
			return fmt.Errorf("tcp: read: %w", err)
		}
		if !bytes.Equal(recv, send) {
			return fmt.Errorf("tcp: echo mismatch")
		}
		return nil
	}

	// Warmup (discarded).
	for i := 0; i < cfg.Warmup; i++ {
		if err := roundTrip(); err != nil {
			return nil, err
		}
	}

	// Measured.
	samples := make([]int64, cfg.Iterations)
	for i := 0; i < cfg.Iterations; i++ {
		start := time.Now()
		if err := roundTrip(); err != nil {
			return nil, err
		}
		samples[i] = time.Since(start).Nanoseconds()
	}

	// Close the client so the server's echo loop sees EOF and exits cleanly,
	// then surface any server-side error that isn't a normal close.
	conn.Close()
	ln.Close()
	if err := <-srvErr; err != nil {
		return nil, err
	}
	return samples, nil
}

// tcpEchoServer accepts one connection and echoes fixed-size payloads back
// until the client disconnects.
func tcpEchoServer(ln net.Listener, payloadBytes int, errc chan<- error) {
	conn, err := ln.Accept()
	if err != nil {
		errc <- nil // listener closed; not an error worth reporting
		return
	}
	defer conn.Close()

	buf := make([]byte, payloadBytes)
	for {
		if _, err := io.ReadFull(conn, buf); err != nil {
			if err == io.EOF || err == io.ErrUnexpectedEOF {
				errc <- nil
				return
			}
			errc <- fmt.Errorf("tcp echo: read: %w", err)
			return
		}
		if _, err := conn.Write(buf); err != nil {
			errc <- fmt.Errorf("tcp echo: write: %w", err)
			return
		}
	}
}

File: rust/network-rtt/src/tcp.rs
//! TCP synchronous ping-pong RTT measurement over loopback.
//!
//! An in-process echo server runs on its own thread; a single client connection
//! issues one request at a time and times each round trip.

use std::io::{self, Read, Write};
use std::net::{TcpListener, TcpStream};
use std::thread;
use std::time::Instant;

use crate::config::Config;

/// Run the TCP benchmark and return one elapsed-nanosecond sample per measured
/// round trip (`cfg.iterations` of them).
pub fn run(cfg: &Config) -> io::Result<Vec<u64>> {
    // Bind to an ephemeral loopback port so the OS picks a free one.
    let listener = TcpListener::bind("127.0.0.1:0")?;
    let addr = listener.local_addr()?;
    let payload_bytes = cfg.payload_bytes;

    // Echo server: accept one connection, echo bytes until the client closes.
    let server = thread::spawn(move || -> io::Result<()> {
        let (mut conn, _) = listener.accept()?;
        conn.set_nodelay(true)?;
        let mut buf = vec![0u8; payload_bytes];
        loop {
            // Read exactly one payload, echo it back. EOF ends the loop.
            match read_exact_or_eof(&mut conn, &mut buf)? {
                false => return Ok(()), // clean EOF
                true => conn.write_all(&buf)?,
            }
        }
    });

    let result = client_loop(addr, cfg);

    // Closing the client stream (dropped inside client_loop) yields EOF on the
    // server side; join surfaces any server-side IO error.
    server
        .join()
        .map_err(|_| io::Error::new(io::ErrorKind::Other, "tcp echo server thread panicked"))??;

    result
}

/// Connect, warm up, then measure `cfg.iterations` round trips.
fn client_loop(addr: std::net::SocketAddr, cfg: &Config) -> io::Result<Vec<u64>> {
    let mut stream = TcpStream::connect(addr)?;
    stream.set_nodelay(true)?;

    let send = vec![0xABu8; cfg.payload_bytes];
    let mut recv = vec![0u8; cfg.payload_bytes];

    // Warmup — timings discarded.
    for _ in 0..cfg.warmup {
        round_trip(&mut stream, &send, &mut recv)?;
    }

    // Pre-allocate the sample buffer so allocation never enters the timed path.
    let mut samples = vec![0u64; cfg.iterations];
    for slot in samples.iter_mut() {
        let start = Instant::now();
        round_trip(&mut stream, &send, &mut recv)?;
        *slot = start.elapsed().as_nanos() as u64;
    }

    Ok(samples)
}

/// One ping-pong: write the full payload, read the full echo, assert equality.
#[inline]
fn round_trip(stream: &mut TcpStream, send: &[u8], recv: &mut [u8]) -> io::Result<()> {
    stream.write_all(send)?;
    stream.read_exact(recv)?;
    if recv != send {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "tcp echo mismatch: received bytes differ from sent",
        ));
    }
    Ok(())
}

/// Fill `buf` from `conn`; returns `Ok(true)` if a full buffer was read,
/// `Ok(false)` on clean EOF before any bytes, `Err` on short EOF or IO error.
fn read_exact_or_eof(conn: &mut TcpStream, buf: &mut [u8]) -> io::Result<bool> {
    let mut read = 0;
    while read < buf.len() {
        match conn.read(&mut buf[read..]) {
            Ok(0) => {
                if read == 0 {
                    return Ok(false); // clean EOF at a payload boundary
                }
                return Err(io::Error::new(
                    io::ErrorKind::UnexpectedEof,
                    "tcp echo server: connection closed mid-payload",
                ));
            }
            Ok(n) => read += n,
            Err(e) if e.kind() == io::ErrorKind::Interrupted => continue,
            Err(e) => return Err(e),
        }
    }
    Ok(true)
}
